### ISSUE TITLE
Strange behaviour of cv2.add operation

### ISSUE THREAD

**Issue by on2k23nm** [opencv/opencv#25165]
### System Information

OpenCV python version: 4.9.0.80
Operating System / Platform: Ubuntu 22.04.4 LTS (WSL)
Python version: 3.10.12

### Detailed description

I am expecting that cv2.add will truncate the results of addition between [0, 255], but that's not happening in cv2 4.9.0.80 as can be seen below :
```
(py2k23) onkar@DESKTOP-MT8AVE7:~/$ python
Python 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> x = np.uint8([250])
>>> y = np.uint8([10])
>>> import cv2
>>> print(cv2.add(x, y))
[[260.]
 [  0.]
 [  0.]
 [  0.]]
>>> print(x+y)
[4]
>>> print(cv2.__version__)
4.9.0
>>> 
```

I tested the same behavior in other environment  which is having cv2 3.4.18.65 and there it is working as expected.

```
(myPyTestEnv) onkar@DESKTOP-MT8AVE7:~$ python
Python 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> import cv2
>>> x = np.uint8([250])
>>> y = np.uint8([10])
>>> z = cv2.add(x, y)
>>> print(z)
[[255]]
>>> print(cv2.__version__)
3.4.18
>>>
```

### Steps to reproduce

To repro the problem, just follow the OpenCV tutorials [here](https://docs.opencv.org/3.4/d0/d86/tutorial_py_image_arithmetics.html) with cv2 version 4.9.0.80

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

**Comment by Garvanand** [comment]
### Check if there are any known compatibility issues with OpenCV version 4.9.0.80 on Ubuntu 22.04.4 LTS (WSL). The issue may be platform-specific.(Try this first)

 If the issue is critical for your work, you may consider downgrading OpenCV to version 3.4.18.65 as a temporary solution until the problem is resolved in a future release.

If the issue is Open, kindly assign the issue to me so that I can work upon it and dig deeper to find the exact issue and the root cause.

**Comment by on2k23nm** [comment]
> ### Check if there are any known compatibility issues with OpenCV version 4.9.0.80 on Ubuntu 22.04.4 LTS (WSL). The issue may be platform-specific.(Try this first)
> If the issue is critical for your work, you may consider downgrading OpenCV to version 3.4.18.65 as a temporary solution until the problem is resolved in a future release.
> 
> If the issue is Open, kindly assign the issue to me so that I can work upon it and dig deeper to find the exact issue and the root cause.

1) I have already moved on with 3.4.18.65 cv2 release
2) What compability issues are you referring to. I'm running in venv and not directly in WSL. So I don't think this is a compatibility issue. 

Anyways, someone has to try this out and ascertain if this is really a issue or not. Currently I don't have time to look into this otherwise I could have digged in deeper into this.

**Comment by LaurentBerger** [comment]
As x and y are not 2 or 3 dimensional array opencv python binding 

y = 10*np.ones((1,1,1), np.uint8)
x = 250*np.ones((1,1,1), np.uint8)
cv.add(x, y)

**Comment by Kumataro** [comment]
Maybe (it is not investigated) caused by https://github.com/opencv/opencv/pull/24074

**Comment by crackwitz** [comment]
the Stack Overflow questions that those people came from:
* https://stackoverflow.com/questions/77897622/opencv-add-operation-results-in-unexpected-dtype-and-shape
* https://stackoverflow.com/questions/78111233/strange-behaviour-of-cv2-add-operation

**Comment by Kumataro** [comment]
I challenged to make patch to avoid convertion single long/double argument to cv::Mat for add/subtract/...

Currently python binding doesn't support changing argument-types. 
`add()` function can be used for src1/src2 with only `Mat`  , not `Scalar`, `int` and `double`.
It cause some strange behavious with arithm functions for emulation  other types.

This patch is able to use manual some implementation functions instead of auto-generation functions.
So we can modify binding functions as we want to.

However, this brute-force fix may not be suitable to maintainance for us.
If there is a better way to fix this, I would like to use it.

**Comment by knmcguire** [comment]
Hi! I saw this too in the python-bindings of 4.10  when going through the image arithmetic documentation

Is it perhaps also an idea to add the double bracket solution that @crackwitz proposed in stackoverflow to the basic documentation/tutorial like here? 

https://github.com/opencv/opencv/blob/78630ddc678c7213ef84aa5a12650d3f22c06590/doc/py_tutorials/py_core/py_image_arithmetics/py_image_arithmetics.markdown?plain=1#L22-L23

**Comment by crackwitz** [comment]
I'd recommend using `.reshape()`...

... and to *never ever* say "double bracket". That wording misses the point (three dimensions) and may misguide newbies.

**Comment by knmcguire** [comment]
Yes, good catch. I meant nested brackets, but double is a bit informal. Thanks for the feedback albeit a bit harsh.

But reshape is cleaner I agree.  It's a bit too bad for a simple line of code that that is necessary but at least people know more what is going on. Python bindings are messy

**Comment by asmorkalov** [comment]
I found the issue reason. Python bindings convert arrays with less than or equal  4 elements to cv::Scalar here: https://github.com/opencv/opencv/blob/c5860d0d752f1cece797765e42f73ca7085333e3/modules/python/src2/cv2_convert.cpp#L276. It was done intentionally to fix https://github.com/opencv/opencv/issues/24057. Looking on more accurate fix.
